apps/www/_blog/2023-08-11-supavisor-1-million.mdx

```diff
@@ -0,0 +1,436 @@
+
+<div>
+  <img
+    alt="multi database"
+    className="dark:hidden"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-light.png"
+  />
+  <img
+    alt="multi database"
+    className="hidden dark:block"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-dark.png"
+  />
+</div>
+
+Supavisor will enable us to build some exciting new features:
+
+- query caching
+- automatic read-replica load balancing
+- query blocking
+- and much more
+
+## Benchmarking 1 million connections
+
+Before rolling out Supavisor to our entire Postgres fleet, we wanted to benchmark the characteristics it exhibited under load. We tested how we can scale the cluster to handle workloads both vertically and horizontally on demand. These results have given us confidence that Supavisor can meet the needs of the Supabase platform.
+
+### Setup
+
+We use a [custom load-testing application](https://github.com/supabase/benchmarks) to test the features of the Supabase platform. It consists of:
+
+- Terraform scripts for creating a testing environment on AWS.
+- k6 as the load generator. We used the k6 guides for [running large-scale tests](https://k6.io/docs/testing-guides/running-large-tests/) and [fine-tuning OS](https://k6.io/docs/misc/fine-tuning-os/) to tweak the config for AWS instances.
+- Grafana + Prometheus for monitoring.
+
+To simulate 1,000,000 concurrent active connections, we used 20 AWS EC2 instances with 16 cores and 32GB of RAM. We ran the tests for up to 2 hours to ensure that the Supavisor could handle the load over long periods.
+
+### Establishing a baseline
+
+In the first test, we set up a single ARM 16-core Supavisor instance on Ubuntu 22.04.2 aarch64 connected to one database instance.
+
+We wanted to assess the capacity of a single Supavisor instance. We achieved:
+
+- 250,000 concurrent connections to Supavisor
+- Supavisor was running with a pool of 400 direct connections to the database
+- The system was processing 20,000 queries per second (QPS)
+
+<div>
+  <img
+    alt="diagram supavisor 1 instance light"
+    className="dark:hidden"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-baseline-light.png"
+  />
+  <img
+    alt="diagram supavisor 1 instance dark"
+    className="hidden dark:block"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-baseline-dark.png"
+  />
+</div>
+
+At this point, the database was the bottleneck - 20,000 QPS was the maximum this instance could handle. Increasing QPS is possible with a larger instance or read-replicas, but wanted to focus on the scalability of Supavisor's connection limit. As Supavisor is multitenant and built with Elixir, scaling up is as easy as adding a new tenant.
```

Comment:
```suggestion
With this setup the database is the bottleneck - 20,000 QPS was the maximum this instance could handle. Increasing QPS would have been possible with a larger instance or read-replicas, but wanted to focus on the scalability of Supavisor's connection limit (not Postgres's). Since Supavisor is built with Elixir, scaling is as easy as adding a new tenant.
```

Reply:
```suggestion
With this setup the database is the bottleneck - 20,000 QPS was the maximum this instance could handle. Increasing QPS would have been possible with a larger instance or read-replicas, but we wanted to focus on the scalability of Supavisor's connection limit (not Postgres's). Since Supavisor is built with multitenancy, addition of read-replicas is as easy as sending a single post request.
```